optimism/client_blocks_bedrock_tracers.go

```diff
@@ -0,0 +1,149 @@
+package optimism
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/ethereum-optimism/optimism/l2geth/common"
+	L2Eth "github.com/ethereum-optimism/optimism/l2geth/eth"
+	"github.com/ethereum-optimism/optimism/l2geth/rpc"
+	EthCommon "github.com/ethereum/go-ethereum/common"
+)
+
+// geth traces types
+type rpcCall struct {
+	Result *Call `json:"result"`
+}
+
+// TraceTransactions returns traces for each of the given transactions.
+// TraceTransactions uses `debug_traceTransaction` under the hood.
+func (ec *Client) TraceTransactions(
+	ctx context.Context,
+	blockHash EthCommon.Hash,
+	txs []BedrockRPCTransaction,
+) (map[string][]*FlatCall, error) {
+	if err := ec.traceSemaphore.Acquire(ctx, semaphoreTraceWeight); err != nil {
+		return nil, err
+	}
+	defer ec.traceSemaphore.Release(semaphoreTraceWeight)
+
+	m := make(map[string][]*FlatCall)
+	traces := make([]*Call, len(txs))
+	if len(txs) == 0 {
+		return m, nil
+	}
+
+	// Check the trace cache
+	if ec.traceCache != nil {
+		for i := range txs {
+			result, err := ec.traceCache.FetchTransaction(ctx, common.Hash(*txs[i].TxHash))
+			if err != nil {
+				return nil, err
+			}
+			traces[i] = result
+		}
+		return m, nil
```

Comment:
This doesn't work if the trace cache is enabled because an empty map is returned.

Reply:
fixed, great catch 💯